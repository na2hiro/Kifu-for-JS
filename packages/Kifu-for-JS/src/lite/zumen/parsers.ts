/**
 * This parser is deprecated and kept for documentation purposes.
 * TODO: rewrite so it will generate object which Zumen component accepts.
 */
const HanSuuji = "123456789";
export const ZenSuuji = "１２３４５６７８９";
export const KanSuuji = "一二三四五六七八九";
export const KomaStr = "玉飛角金銀桂香歩玉龍馬金全圭杏と";
const convMgRepeat = function (str) {
    let re = new RegExp("[" + ZenSuuji + KanSuuji + "]", "g");
    str = str.replace(re, function (a) {
        return HanSuuji.charAt((ZenSuuji + KanSuuji).indexOf(a) % 9);
    });
    str = str.replace("０", "0");
    str = str.replace(/十([1-9])/, "1$1");
    str = str.replace("十", "10");
    str = str.replace(/(.)(1?[0-9])/g, function (a, b, c) {
        for (let i = 0, s = ""; i < c; i++) {
            s += b;
        }
        return s;
    });
    return str;
};
const convKomaStr = function (str) {
    str = str.replace("王", "玉");
    str = str.replace("竜", "龍");
    str = str.replace("仝", "杏");
    str = str.replace("个", "と");
    return str;
};
/**
 * Converts the given text to a shogi board representation.
 *
 * Ban is an array of 98 elements.
 * <ul>
 * <li>The first 81 elements (indexed from 0 to 80) are the board squares, starting from the top left.<ul>
 *   <li>5th position: turn</li>
 *   <li>4th position: promoted or not</li>
 * </ul>3rd to 1st: kind</li>
 * <li>The next 7 elements (indexed from 81 to 87) are the amount of captured pieces for the first player.</li>
 * <li>The next 7 elements (indexed from 88 to 93) are the amount of captured pieces for the second player.</li>
 * <li>94 ?</li>
 * <li>The next 2 elements (indexed from 95 to 96) are name of the players.</li>
 * <li>The last element (indexed 97) is the position of the latest move. negative if not available.</li>
 * </ul>
 *
 * @param {string} text - The text to parse.
 * @returns {number[]} - The shogi board representation.
 */
export const readzumen = function (text: string) {
    let mr, mk, x, y, s, c, i, re2;
    const ban = new Array(98);
    for (let i = 0; i < 95; i++) {
        ban[i] = 0;
    }
    ban[95] = "先手";
    ban[96] = "後手";
    ban[97] = -1;
    let mgStrSente = "";
    let mgStrGote = "";
    let re = /^([先下後上]手の)?持駒：([^\n]*)/gm;
    while ((mr = re.exec(text)) != null) {
        if (/[後上]/.test(mr[1])) {
            mgStrGote = mr[2];
        } else {
            mgStrSente = mr[2];
        }
    }
    re = /^([先下後上]手)：(.*)\n/gm;
    while ((mr = re.exec(text)) != null) {
        mr[2].replace(/^\s+/, "");
        mr[2].replace(/(\s|　).*$/, "");
        if (/[後上]/.test(mr[1])) {
            ban[96] = mr[2];
        } else {
            ban[95] = mr[2];
        }
    }
    y = 1;
    re = /^\|([^\n]*)/gm;
    while ((mr = re.exec(text)) != null) {
        mr[1] = convKomaStr(mr[1]);
        x = 9;
        re2 = new RegExp("([\\+\\-vV\\^\\s])(.)", "g");
        while ((mk = re2.exec(mr[1])) != null) {
            s = "vV-".indexOf(mk[1]) > -1 ? 2 : 0; // 5th: turn
            if ((c = KomaStr.indexOf(mk[2])) > -1) {
                if (c > 7) {
                    c -= 8;
                    s = s | 1; // 4th: promoted
                }
                // xysAry[c]+=""+x+y+s;
                ban[y * 9 - x] = s * 8 + c + 1; // 1st-3rd: kind
            }
            if (--x < 1) {
                break;
            }
        }
        if (++y > 9) {
            break;
        }
    }
    if (y == 1) {
        return false;
    }
    mgStrSente = convMgRepeat(mgStrSente);
    for (i = 0; i < mgStrSente.length; i++) {
        if ((c = KomaStr.indexOf(mgStrSente.charAt(i))) > 0) {
            ban[81 + c - 1] += 1;
        }
    }
    mgStrGote = convMgRepeat(mgStrGote);
    for (i = 0; i < mgStrGote.length; i++) {
        if ((c = KomaStr.indexOf(mgStrGote.charAt(i))) > 0) {
            ban[88 + c - 1] += 1;
        }
    }
    re = /^手数＝\d+\s+[▲△▽]?(.)(.)(.*)まで/gm;
    if ((mr = re.exec(text)) != null) {
        let mx = -100;
        if ((c = 0 + mr[1]) > 0) {
            mx = 8 - c;
        } else if ((c = ZenSuuji.indexOf(mr[1])) > 0) {
            mx = 8 - c;
        }
        if ((c = 0 + mr[2]) > 0) {
            mx += c * 9;
        } else if ((c = ZenSuuji.indexOf(mr[2])) > 0) {
            mx += c * 9;
        } else if ((c = KanSuuji.indexOf(mr[2])) > 0) {
            mx += c * 9;
        }
        ban[97] = mx;
    }
    return ban;
};
