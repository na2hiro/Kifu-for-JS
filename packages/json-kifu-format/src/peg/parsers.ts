import {IJSONKifuFormat} from "../Formats";
import "./ambient";
import * as CSAParser from "./csa-parser.pegjs";
import * as KI2Parser from "./ki2-parser.pegjs";
import * as KIFParser from "./kif-parser.pegjs";

export function parseCSA(csaString: string): IJSONKifuFormat {
    return CSAParser.parse(csaString);
}

export function parseKIF(kifString: string): IJSONKifuFormat {
    return KIFParser.parse(kifString);
}

export function parseKI2(ki2String: string): IJSONKifuFormat {
    return KI2Parser.parse(ki2String);
}
