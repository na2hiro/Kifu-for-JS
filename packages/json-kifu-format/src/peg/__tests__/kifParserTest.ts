/* eslint-disable no-irregular-whitespace */
import {parseKIF as parse} from "../parsers";

function p(x, y) {
    return {x, y};
}
describe("kif-parser", () => {
    it("simple", () => {
        expect(
            parse("1 ７六歩(77)\n2 ３四歩(33)\n3 ２二角成(88)\n 4 同　銀(31)\n5 ４五角打\n")
        ).toEqual({
            header: {},
            moves: [
                {},
                {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                {move: {from: p(3, 1), same: true, piece: "GI"}},
                {move: {to: p(4, 5), piece: "KA"}},
            ],
        });
    });
    it("comment", () => {
        expect(
            parse(
                "*開始時コメント\n1 ７六歩(77)\n*初手コメント\n*初手コメント2\n2 ３四歩(33)\n3 ２二角成(88)\n"
            )
        ).toEqual({
            header: {},
            moves: [
                {comments: ["開始時コメント"]},
                {
                    move: {from: p(7, 7), to: p(7, 6), piece: "FU"},
                    comments: ["初手コメント", "初手コメント2"],
                },
                {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
            ],
        });
    });
    it("time", () => {
        expect(
            parse(
                "1 ７六歩(77) (0:01/00:00:01)\n2 ３四歩(33) (0:02/00:00:02)\n3 ２二角成(88) (0:20/00:00:21)\n 4 同　銀(31) (0:03/00:00:05)\n5 ４五角打 (0:39/00:01:00)\n"
            )
        ).toEqual({
            header: {},
            moves: [
                {},
                {
                    move: {from: p(7, 7), to: p(7, 6), piece: "FU"},
                    time: {now: {m: 0, s: 1}, total: {h: 0, m: 0, s: 1}},
                },
                {
                    move: {from: p(3, 3), to: p(3, 4), piece: "FU"},
                    time: {now: {m: 0, s: 2}, total: {h: 0, m: 0, s: 2}},
                },
                {
                    move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true},
                    time: {now: {m: 0, s: 20}, total: {h: 0, m: 0, s: 21}},
                },
                {
                    move: {from: p(3, 1), same: true, piece: "GI"},
                    time: {now: {m: 0, s: 3}, total: {h: 0, m: 0, s: 5}},
                },
                {
                    move: {to: p(4, 5), piece: "KA"},
                    time: {now: {m: 0, s: 39}, total: {h: 0, m: 1, s: 0}},
                },
            ],
        });
    });
    it("special", () => {
        expect(
            parse(
                "1 ７六歩(77)\n2 ３四歩(33)\n3 ７八銀(79)\n 4 ８八角成(22)\n5 投了\nまで4手で後手の勝ち\n"
            )
        ).toEqual({
            header: {},
            moves: [
                {},
                {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                {move: {from: p(7, 9), to: p(7, 8), piece: "GI"}},
                {move: {from: p(2, 2), to: p(8, 8), piece: "KA", promote: true}},
                {special: "TORYO"},
            ],
        });
    });
    describe("header", () => {
        it("手合割", () => {
            expect(
                parse(
                    "手合割：平手\n1 ７六歩(77)\n2 ３四歩(33)\n3 ２二角成(88)\n 4 同　銀(31)\n5 ４五角打\n"
                )
            ).toEqual({
                header: {
                    手合割: "平手",
                },
                initial: {preset: "HIRATE"},
                moves: [
                    {},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                    {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                    {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                    {move: {from: p(3, 1), same: true, piece: "GI"}},
                    {move: {to: p(4, 5), piece: "KA"}},
                ],
            });
            expect(
                parse(
                    "手合割：六枚落ち\n1 ４二玉(51)\n2 ７六歩(77)\n3 ２二銀(31)\n 4 ６六角(88)\n5 ８二銀(71)\n"
                )
            ).toEqual({
                header: {
                    手合割: "六枚落ち",
                },
                initial: {preset: "6"},
                moves: [
                    {},
                    {move: {from: p(5, 1), to: p(4, 2), piece: "OU"}},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                    {move: {from: p(3, 1), to: p(2, 2), piece: "GI"}},
                    {move: {from: p(8, 8), to: p(6, 6), piece: "KA"}},
                    {move: {from: p(7, 1), to: p(8, 2), piece: "GI"}},
                ],
            });
        });
    });
    describe("initial", () => {
        it("simple", () => {
            expect(
                parse(
                    "\
手合割：その他　\n\
上手の持駒：銀四　桂四　\n\
  ９ ８ ７ ６ ５ ４ ３ ２ １\n\
+---------------------------+\n\
| ・ ・ ・ ・ ・ ・ ・v歩v玉|一\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|二\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|三\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|四\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|五\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|六\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|七\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|八\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|九\n\
+---------------------------+\n\
下手の持駒：飛二　香四　\n\
下手番\n\
下手：shitate\n\
上手：uwate\n\
1 １三香打\n2 １二桂打\n3 同　香成(13)\n"
                )
            ).toEqual({
                header: {
                    手合割: "その他　",
                    上手: "uwate",
                    下手: "shitate",
                },
                initial: {
                    preset: "OTHER",
                    data: {
                        board: [
                            [{color: 1, kind: "OU"}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                            ],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                        ],
                        color: 0,
                        hands: [
                            {FU: 0, KY: 4, KE: 0, GI: 0, KI: 0, KA: 0, HI: 2},
                            {FU: 0, KY: 0, KE: 4, GI: 4, KI: 0, KA: 0, HI: 0},
                        ],
                    },
                },
                moves: [
                    {},
                    {move: {to: p(1, 3), piece: "KY"}},
                    {move: {to: p(1, 2), piece: "KE"}},
                    {move: {from: p(1, 3), same: true, piece: "KY", promote: true}},
                ],
            });
        });
        it("Kifu for iPhone dialect", () => {
            // 持ち駒が半角スペース区切り
            // 手合割が平手
            expect(
                parse(
                    "\
手合割：平手\n\
上手の持駒：銀四 桂四 \n\
  ９ ８ ７ ６ ５ ４ ３ ２ １\n\
+---------------------------+\n\
| ・ ・ ・ ・ ・ ・ ・v歩v玉|一\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|二\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|三\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|四\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|五\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|六\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|七\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|八\n\
| ・ ・ ・ ・ ・ ・ ・v歩 ・|九\n\
+---------------------------+\n\
下手の持駒：飛二 香四 \n\
下手番\n\
下手：shitate\n\
上手：uwate\n\
1 １三香打\n2 １二桂打\n3 同　香成(13)\n"
                )
            ).toEqual({
                header: {
                    手合割: "平手",
                    上手: "uwate",
                    下手: "shitate",
                },
                initial: {
                    preset: "OTHER", // 上書き
                    data: {
                        board: [
                            [{color: 1, kind: "OU"}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                                {color: 1, kind: "FU"},
                            ],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                            [{}, {}, {}, {}, {}, {}, {}, {}, {}],
                        ],
                        color: 0,
                        hands: [
                            {FU: 0, KY: 4, KE: 0, GI: 0, KI: 0, KA: 0, HI: 2},
                            {FU: 0, KY: 0, KE: 4, GI: 4, KI: 0, KA: 0, HI: 0},
                        ],
                    },
                },
                moves: [
                    {},
                    {move: {to: p(1, 3), piece: "KY"}},
                    {move: {to: p(1, 2), piece: "KE"}},
                    {move: {from: p(1, 3), same: true, piece: "KY", promote: true}},
                ],
            });
        });
    });
    describe("fork", () => {
        it("normal", () => {
            expect(
                parse(
                    "\
手合割：平手\n\
1 ７六歩(77)\n\
2 ３四歩(33)\n\
3 ２二角成(88)+\n\
4 同　銀(31)\n\
5 ４五角打\n\
6 中断\n\
\n\
変化：3手\n\
3 ６六歩(67)\n\
4 ８四歩(83)\n\
"
                )
            ).toEqual({
                header: {
                    手合割: "平手",
                },
                initial: {preset: "HIRATE"},
                moves: [
                    {},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                    {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                    {
                        move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true},
                        forks: [
                            [
                                //                            {},
                                {move: {from: p(6, 7), to: p(6, 6), piece: "FU"}},
                                {move: {from: p(8, 3), to: p(8, 4), piece: "FU"}},
                            ],
                        ],
                    },
                    {move: {from: p(3, 1), same: true, piece: "GI"}},
                    {move: {to: p(4, 5), piece: "KA"}},
                    {special: "CHUDAN"},
                ],
            });
        });

        it("Multiple forks", () => {
            expect(
                parse(
                    "1 ７六歩(77)+\n" +
                        "2 ３四歩(33)+\n" +
                        "\n" +
                        "変化：2手\n" +
                        "2 ８四歩(83)+\n" +
                        "\n" +
                        "変化：2手\n" +
                        "2 ４四歩(43)\n" +
                        "\n" +
                        "変化：1手\n" +
                        "1 ２六歩(27)+\n" +
                        "\n" +
                        "変化：1手\n" +
                        "1 ８六歩(87)+\n" +
                        "\n" +
                        "変化：1手\n" +
                        "1 ７八金(69)\n"
                )
            ).toEqual({
                header: {},
                moves: [
                    {},
                    {
                        move: {from: p(7, 7), to: p(7, 6), piece: "FU"},
                        forks: [
                            [{move: {from: p(2, 7), to: p(2, 6), piece: "FU"}}],
                            [{move: {from: p(8, 7), to: p(8, 6), piece: "FU"}}],
                            [{move: {from: p(6, 9), to: p(7, 8), piece: "KI"}}],
                        ],
                    },
                    {
                        move: {from: p(3, 3), to: p(3, 4), piece: "FU"},
                        forks: [
                            [{move: {from: p(8, 3), to: p(8, 4), piece: "FU"}}],
                            [{move: {from: p(4, 3), to: p(4, 4), piece: "FU"}}],
                        ],
                    },
                ],
            });
        });
    });
    describe("split", () => {
        it("normal", () => {
            expect(
                parse(
                    "\
手合割：平手\n\
手数----指手--\n\
*開始コメント\n\
1 ７六歩(77)\n\
*初手コメント\n\
2 ３四歩(33)\n\
3 ２二角成(88)+\n\
4 中断\n\
"
                )
            ).toEqual({
                header: {
                    手合割: "平手",
                },
                initial: {preset: "HIRATE"},
                moves: [
                    {comments: ["開始コメント"]},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}, comments: ["初手コメント"]},
                    {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                    {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                    {special: "CHUDAN"},
                ],
            });
        });
        it("after initial comment", () => {
            expect(
                parse(
                    "\
手合割：平手\n\
*開始コメント\n\
手数----指手--\n\
1 ７六歩(77)\n\
*初手コメント\n\
2 ３四歩(33)\n\
3 ２二角成(88)+\n\
4 中断\n\
"
                )
            ).toEqual({
                header: {
                    手合割: "平手",
                },
                initial: {preset: "HIRATE"},
                moves: [
                    {comments: ["開始コメント"]},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}, comments: ["初手コメント"]},
                    {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                    {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                    {special: "CHUDAN"},
                ],
            });
        });
    });
    describe("unsupported annotations", () => {
        it("盤面回転", () => {
            expect(
                parse(
                    "盤面回転\n1 ７六歩(77)\n2 ３四歩(33)\n3 ２二角成(88)\n 4 同　銀(31)\n5 ４五角打\n"
                )
            ).toEqual({
                header: {},
                moves: [
                    {},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                    {move: {from: p(3, 3), to: p(3, 4), piece: "FU"}},
                    {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                    {move: {from: p(3, 1), same: true, piece: "GI"}},
                    {move: {to: p(4, 5), piece: "KA"}},
                ],
            });
        });
        it("&読み込み時表示", () => {
            expect(
                parse(
                    "1 ７六歩(77)\n2 ３四歩(33)\n&読み込み時表示\n3 ２二角成(88)\n 4 同　銀(31)\n5 ４五角打\n"
                )
            ).toEqual({
                header: {},
                moves: [
                    {},
                    {move: {from: p(7, 7), to: p(7, 6), piece: "FU"}},
                    {
                        move: {from: p(3, 3), to: p(3, 4), piece: "FU"},
                        comments: ["&読み込み時表示"],
                    },
                    {move: {from: p(8, 8), to: p(2, 2), piece: "KA", promote: true}},
                    {move: {from: p(3, 1), same: true, piece: "GI"}},
                    {move: {to: p(4, 5), piece: "KA"}},
                ],
            });
        });
    });
    it("same at first fork", () => {
        expect(
            parse(`   1 ７六歩(77)   ( 0:00/00:00:00)
   2 ３四歩(33)   ( 0:00/00:00:00)
   3 ２六歩(27)   ( 0:00/00:00:00)
   4 中断         ( 0:00/00:00:00)
まで3手で中断

変化：3手
   3 ２二角成(88) ( 0:00/00:00:00)
   4 同　銀(31)   ( 0:00/00:00:00)+
   5 中断         ( 0:00/00:00:00)
まで4手で中断

変化：4手
   4 同　飛(82)   ( 0:00/00:00:00)
   5 中断         ( 0:00/00:00:00)
まで4手で中断
`)
        ).toMatchSnapshot();
    });
});
