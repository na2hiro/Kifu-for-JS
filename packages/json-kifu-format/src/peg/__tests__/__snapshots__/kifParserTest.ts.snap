// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kif-parser same at first fork 1`] = `{"header":{},"moves":[{},{"move":{"piece":"FU","to":{"x":7,"y":6},"from":{"x":7,"y":7}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}},{"move":{"piece":"FU","to":{"x":3,"y":4},"from":{"x":3,"y":3}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}},{"move":{"piece":"FU","to":{"x":2,"y":6},"from":{"x":2,"y":7}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}},"forks":[[{"move":{"piece":"KA","to":{"x":2,"y":2},"promote":true,"from":{"x":8,"y":8}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}},{"move":{"piece":"GI","same":true,"from":{"x":3,"y":1}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}},"forks":[[{"move":{"piece":"HI","same":true,"from":{"x":8,"y":2}},"time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}},{"special":"CHUDAN","time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}}]]},{"special":"CHUDAN","time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}}]]},{"special":"CHUDAN","time":{"now":{"m":0,"s":0},"total":{"h":0,"m":0,"s":0}}}]}`;
